packages/common/src/AncillaryDataParser.ts

```diff
@@ -0,0 +1,255 @@
+  keyValues.forEach((keyValue: any) => {
+    const [key, value] = parseKeyValue(keyValue);
+    ancillaryObject[key] = value;
+  });
+  return ancillaryObject;
+}
+
+// Escapes double quoted keys/values and values enclosed in curly/square brackets.
+function markEscapes(stringObject: any) {
+  stringObject.forEach((charObject: any, openIndex: number, stringObject: any) => {
+    // Skip searching in already escaped characters or closing double quotes:
+    if (charObject.escape || charObject.skip) return;
+
+    // Escape keys: opening quotes should be after comma (,) separator or start.
+    if (
+      charObject.character == '"' &&
+      (isNextEnd(stringObject, openIndex, false) || isNextChar(stringObject, openIndex, ",", false))
+    )
+      escapeQuotes(stringObject, openIndex, false);
+
+    // Escape string values: opening quotes should be after column (:) separator.
+    if (charObject.character == '"' && isNextChar(stringObject, openIndex, ":", false))
+      escapeQuotes(stringObject, openIndex);
+
+    // Escape JSON values: first opening curly brackets should be after column (:) separator.
+    if (charObject.character == "{" && isNextChar(stringObject, openIndex, ":", false))
+      escapeJSON(stringObject, openIndex);
+
+    // Escape JSON values: first opening square brackets should be after column (:) separator.
+    if (charObject.character == "[" && isNextChar(stringObject, openIndex, ":", false))
+      escapeJSON(stringObject, openIndex, false);
+  });
+}
+
+// Splits ancillary data object into key-value pairs.
+function splitKeyValues(stringObject: any): any {
+  const keyValues: any = [];
+  for (let startIndex = 0; startIndex < stringObject.length; startIndex++) {
+    const charObject: any = stringObject[startIndex];
+
+    // If reached unescaped comma (,) continue with the next key-value pair:
+    if (!skipWhitespace(charObject) || (charObject.character == "," && !charObject.escape)) continue;
+
+    for (let endIndex = startIndex; endIndex < stringObject.length; endIndex++) {
+      // Search for next unescaped comma (,) delimiter or end of object:
+      if (
+        endIndex == stringObject.length - 1 ||
+        isNextEnd(stringObject, endIndex) ||
+        isNextChar(stringObject, endIndex, ",")
+      ) {
+        // Copy the identified key-value pair:
+        const pairIndex = keyValues.length;
+        keyValues.push([]);
+        for (let i = startIndex; i <= endIndex; i++) {
+          keyValues[pairIndex].push(stringObject[i]);
+        }
+
+        // Skip start index to the end of current key-value pair:
+        startIndex = endIndex;
+        break;
+      }
+    }
+  }
+
+  // Remove enclosing double quotes.
+  keyValues.forEach((keyValue: any, index: number, keyValues: any) => {
+    keyValues[index] = keyValue.filter(removeDoubleQuotes);
+  });
+  return keyValues;
+}
+
+// Tries to parse key:value pair.
+function parseKeyValue(keyValue: any): any {
```

Comment:
nit: can we make these types more specific?

Reply:
I'm not sure in this case, since all of the following are valid key-value pairs:

`{ key: 'value' }`
`{ key1: 'value1', key2: 'value2' }`
`{ '  key ': '  value  ' }`
`{ '"  key': 'value  "' }`
`{ key: { nestedKey: 'nestedValue' } }`
`{ key: '{nestedKey: "nestedValue",}' }`
`{ key: [ 1, 'a', null, true, {}, [] ] }`

Testing for the specific disallowed cases (key with no value, empty key, multiple column delimiters in key-value pairs, JSON object in the key, random hex) might be better than trying to strictly define what _is_ allowed through types?

Reply:
Thought a bit more, and it appears `keyValue` should be `CharObject[]` and `keyValues` should be `[CharObject[]]`. Running with that and seeing how it goes.